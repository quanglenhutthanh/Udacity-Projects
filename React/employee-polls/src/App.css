.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

.question-container {
  max-width: 800px;
  margin: 0 auto;
  /* padding: 20px; */
}

.question-grid {
  list-style: none;
  padding: 0;
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  /* Three columns with equal width */
  gap: 20px;
  /* Adjust the gap between grid items */
  text-align: center;
}

/* Add styles for question details */
.question-details {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-top: 0px;
}

.question-details p {
  font-weight: bold;
  margin: 0; 
  
}

.question-details span {
  font-size: 11px;
  color: #888;
}

/* Adjust the padding and margin for question item */
.question-item {
  border: 2px solid #000;
  padding: 10px;
  margin: 10px; /* Add a margin around each question item */
}


/* Style the button */
.question-item button {
  background-color: #007bff;
  color: #fff;
  border: none;
  padding: 8px 12px;
  cursor: pointer;
  font-size: 14px;
  margin-top: 10px;
}

.question-item button:hover {
  background-color: #0056b3;
}

.login-container {
  
  display: flex;
  align-items: center;
  justify-content: center;
  
}

.login-container h2 {
  color: #333;
  margin-bottom: 20px; /* Added margin for spacing between heading and form */
  
}

.login-form {
  display: flex;
  flex-direction: column;
  margin-top: 20px;
}

.form-group {
  margin-bottom: 15px;
}

.form-group label {
  
  font-weight: bold;
  margin-bottom: 5px;
}

.form-group input {
  width: 100%;
  padding: 8px;
  box-sizing: border-box;
}

.login-btn {
  background-color: #007bff;
  color: #fff;
  border: none;
  padding: 10px;
  cursor: pointer;
}

.login-btn:hover {
  background-color: #0056b3;
}