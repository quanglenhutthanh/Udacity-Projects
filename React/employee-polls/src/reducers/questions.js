import { RECEIVE_QUESTIONS, ADD_QUESTION, ADD_VOTE } from "../actions/questions";  

export default function questions(state = {}, action){
    switch(action.type)
    {
        case RECEIVE_QUESTIONS:
            console.log(action.questions);
            return{
                ...state,
                ...action.questions
            };
        case ADD_QUESTION:
            return{
                ...state,
                [action.question.id]: action.question,
            };
            case ADD_VOTE:
                return {
                  ...state,
                  [action.questionId]: {
                    ...state[action.questionId],
                    [action.vote]: {
                      ...state[action.questionId][action.vote],
                      votes: state[action.questionId][action.vote].votes.concat(action.author)
                    }
                  }
          
          
                };
        default:
            return state;
    }
}