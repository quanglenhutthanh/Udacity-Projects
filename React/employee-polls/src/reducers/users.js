import { RECEIVE_USERS, ADD_QUESTION_USER, ADD_USER_VOTE } from "../actions/users";



export default function users(state = {}, action) {
  switch (action.type) {
    case RECEIVE_USERS:
      console.log(action.users);
      return {
        ...state,
        ...action.users,
      };
    case ADD_QUESTION_USER:
      const authorId = action.author;
      const user = state[authorId.id];

      if (user && user.questions) {
        return {
          ...state,
          [authorId]: {
            ...user,
            questions: user.questions.concat(action.questionId),
          },
        };
      } else {
        console.error(
          `User with ID ${authorId} not found or questions array is undefined.`
        );
        return state;
      }
    case ADD_USER_VOTE:
      return {
        ...state,
        [action.authedUser]: {
          ...state[action.authedUser],
          answers: {
            ...state[action.authedUser].answers,
            [action.questionId]: action.vote
          }
        }
      };
    default:
      return state;
  }
}
