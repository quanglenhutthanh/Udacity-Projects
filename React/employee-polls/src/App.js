import "./App.css";
import { Route, Routes } from "react-router-dom";
import { useEffect } from "react";
import { connect } from "react-redux";
import { handleInitialData } from "./actions/shared";
import Nav from "./components/Nav";
import Home from "./components/Home";
import Login from "./components/Login";
import Poll from "./components/Poll";
import LeaderBoard from "./components/LeaderBoard";
import NewPoll from "./components/NewPoll";
import Page404 from "./components/Page404";
import ProtectedRoute from "./components/ProtectedRoute";

const App = (props) => {
 
  useEffect(() => {
    props.dispatch(handleInitialData());
  });

  return (
    <div>
      {props.loggedIn && <Nav />}
      <Routes>
        <Route path="/" element={<ProtectedRoute><Home/></ProtectedRoute>}/>
        <Route path="/login" element={<Login/>} />
        <Route path="/questions/:id" element={<ProtectedRoute><Poll/></ProtectedRoute>}/>
        <Route path="/add" element={<ProtectedRoute><NewPoll/></ProtectedRoute>}/>
        <Route path="/leaderboard" element={<ProtectedRoute><LeaderBoard/></ProtectedRoute>} />
        <Route path="/404" element={<ProtectedRoute><Page404/></ProtectedRoute>} />
      </Routes>
    </div>
  );
};

const mapStateToProps = ({authedUser}) => ({
  loggedIn: !!authedUser,
});
export default connect(mapStateToProps)(App);
