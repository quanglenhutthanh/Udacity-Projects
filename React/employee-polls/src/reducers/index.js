import questions from "./questions";
import users from "./users";
import authedUser from "./authedUser";
import { combineReducers } from "redux";

export default combineReducers({
    questions,
    users,
    authedUser
});