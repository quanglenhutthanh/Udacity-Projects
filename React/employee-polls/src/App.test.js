import React from 'react';
import {render} from '@testing-library/react';
import App from './App';
import reducers from './reducers';
import {Provider} from "react-redux";
import {configureStore} from '@reduxjs/toolkit';
import {BrowserRouter} from "react-router-dom";


describe("App", () => {
    it("Render App", () => {
        const store = configureStore({reducer:reducers});
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <App/>
                </BrowserRouter>
            </Provider>
        );
        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
    });
});
