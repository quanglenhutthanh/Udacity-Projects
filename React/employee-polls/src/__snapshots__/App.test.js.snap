// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Render App 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          styles="align-items: center;"
        >
          <div
            class="login-container"
          >
            <form
              class="login-form"
            >
              <h2>
                Login
              </h2>
              <div
                class="form-group"
              >
                <label>
                  Username:
                </label>
                <input
                  data-testid="username"
                  id="username"
                  name="username"
                  placeholder="Enter your username"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group"
              >
                <label>
                  Password:
                </label>
                <input
                  data-testid="password"
                  id="password"
                  name="password"
                  placeholder="Enter your password"
                  type="password"
                  value=""
                />
              </div>
              <button
                class="login-btn"
                data-testid="submit"
                type="submit"
              >
                Login
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        styles="align-items: center;"
      >
        <div
          class="login-container"
        >
          <form
            class="login-form"
          >
            <h2>
              Login
            </h2>
            <div
              class="form-group"
            >
              <label>
                Username:
              </label>
              <input
                data-testid="username"
                id="username"
                name="username"
                placeholder="Enter your username"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label>
                Password:
              </label>
              <input
                data-testid="password"
                id="password"
                name="password"
                placeholder="Enter your password"
                type="password"
                value=""
              />
            </div>
            <button
              class="login-btn"
              data-testid="submit"
              type="submit"
            >
              Login
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
